### PR TITLE
Chore: Wallet Refactor

### PR DESCRIPTION
**Pull request by brendan-defi**
**What changed? Why?**
* refactored wallet for walletisland

**Notes to reviewers**

**How has it been tested?**
locally, vitest

### CODE REVIEW

File: src/wallet/components/Wallet.test.tsx [hunk 1]
```diff
@@ -29,45 +29,33 @@ vi.mock('../../ui/react/internal/hooks/useOutsideClick', () => ({
...
     });
 
     render(
       <Wallet>
-        <ConnectWallet />
-        <WalletDropdown />
+        <WalletBasic>
+          <ConnectWallet />
+          <WalletDropdown />
+        </WalletBasic>
```
Comment (alessey) [inline src/wallet/components/Wallet.test.tsx:58]: won't wallet island and and current share the connect wallet components?  would't the only change be <WalletDropdown/> vs. <WalletIsland/>?

File: tailwind.config.js [hunk 2]
```diff
@@ -50,13 +53,55 @@ export default {
...
+          '100%': { opacity: '0', transform: 'translateX(-20px)' },
+        },
+        slideInFromRight: {
+          '0%': { opacity: '0', transform: 'translateX(20px)' },
+          '100%': { opacity: '1', transform: 'translateX(0)' },
+        },
+        slideOutToRight: {
+          '0%': { opacity: '1', transform: 'translateX(0)' },
+          '100%': { opacity: '0', transform: 'translateX(20px)' },
+        },
```
Comment (alessey) [inline tailwind.config.js:83]: possible to break these into parts and compose in animations?

File: src/wallet/components/WalletProvider.tsx [hunk 3]
```diff
@@ -15,11 +22,29 @@ type WalletProviderReact = {
...
-    isOpen,
+    address,
     chain,
+    isOpen,
     setIsOpen,
+    isClosing,
+    setIsClosing,
+    handleClose,
+    containerRef,
   });
```
Comment (dschlabach) [inline src/wallet/components/WalletProvider.tsx:46]: Could you do this on `WalletDropdown` instead? It'd avoid having any mismatches in case we ever update the timing (or allow users to change it themselves)

<div
  className={cn(
    pressable.default,
    color.foreground,
    'absolute right-0 z-10 mt-1 flex w-max min-w-[300px] cursor-default flex-col overflow-hidden rounded-xl',
    isClosing ? 'animate-walletContainerOut' : 'animate-walletContainerIn',
    className,
  )}
  onAnimationEnd={() => {
    if (isClosing) {
      setIsOpen(false);
      setIsClosing(false);
    }
  }}
  data-testid="ockWalletDropdown"
>

File: src/wallet/components/Wallet.tsx [hunk 4]
```diff
@@ -1,45 +1,45 @@
...
+      ) {
+        handleClose();
+      }
     };
-  }, [children]);
+
+    document.addEventListener('click', handleClickOutsideComponent);
+    return () =>
+      document.removeEventListener('click', handleClickOutsideComponent);
+  }, [isOpen, handleClose, walletContainerRef]);
```
Comment (cpcramer) [inline src/wallet/components/Wallet.tsx:30]: Let's use the useClickOutside hook!

File: src/wallet/components/WalletBasic.tsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { Children, useMemo } from 'react';
+import { findComponent } from '../../core-react/internal/utils/findComponent';
+import { cn } from '../../styles/theme';
+import type { WalletReact } from '../types';
+import { ConnectWallet } from './ConnectWallet';
+import { WalletDropdown } from './WalletDropdown';
+import { useWalletContext } from './WalletProvider';
+
+export function WalletBasic({ children, className }: WalletReact) {
```
Comment (cpcramer) [inline src/wallet/components/WalletBasic.tsx:9]: What's the difference between `WalletBasic` and `WalletDefault`?

File: tailwind.config.js [hunk 6]
```diff
@@ -50,13 +53,55 @@ export default {
...
+          '100%': { opacity: '1', transform: 'translateY(0)' },
+        },
+        walletContainerOut: {
+          '0%': { opacity: '1', transform: 'translateY(0)' },
+          '100%': { opacity: '0', transform: 'translateY(-10px)' },
+        },
+        walletIslandContainerItemIn: {
+          '0%': { opacity: '0', transform: 'translateY(10px)' },
+          '100%': { opacity: '1', transform: 'translateY(0)' },
+        },
```
Comment (cpcramer) [inline tailwind.config.js:67]: Since they are specific to `Wallet` and `WalletIsland`, should these configs live elsewhere?

File: tailwind.config.js [hunk 7]
```diff
@@ -50,13 +53,55 @@ export default {
...
+        walletContainerIn: 'walletContainerIn 300ms ease-out',
+        walletContainerOut: 'walletContainerOut 300ms forwards',
+        walletIslandContainerItem1:
+          'walletIslandContainerItemIn 300ms cubic-bezier(0.34, 1.56, 0.64, 1) 100ms forwards',
+        walletIslandContainerItem2:
+          'walletIslandContainerItemIn 300ms cubic-bezier(0.34, 1.56, 0.64, 1) 130ms forwards',
+        walletIslandContainerItem3:
+          'walletIslandContainerItemIn 300ms cubic-bezier(0.34, 1.56, 0.64, 1) 160ms forwards',
+        walletIslandContainerItem4:
+          'walletIslandContainerItemIn 300ms cubic-bezier(0.34, 1.56, 0.64, 1) 190ms forwards',
```
Comment (cpcramer) [inline tailwind.config.js:100]: Since they are specific to Wallet and WalletIsland, should these configs live elsewhere?

File: tailwind.config.js (hunk 2 above)
Comment (alessey) [reply tailwind.config.js:83]: https://github.com/coinbase/onchainkit/pull/1734

File: src/wallet/components/Wallet.test.tsx (hunk 1 above)
Comment (brendan-defi) [reply src/wallet/components/Wallet.test.tsx:58]: yes, that's correct. The reason for this change is that WalletBasic is opinionated about children, destructuring and reorganizing them.

File: src/wallet/components/Wallet.tsx (hunk 4 above)
Comment (brendan-defi) [reply src/wallet/components/Wallet.tsx:30]: oh cool, i think you added this after i made my refactor, super helpful, thanks!

File: src/wallet/components/WalletBasic.tsx (hunk 5 above)
Comment (brendan-defi) [reply src/wallet/components/WalletBasic.tsx:9]: WalletBasic is just a wrapper that destructures children and reorganizes them. 

WalletDefault implements WalletBasic, and looks like this:
image

But WalletBasic can be extended beyond WalletDefault, like this:
image

I'm not wedded to these names, perhaps WalletDefault should become WalletBasicDefault or something.

File: src/wallet/components/WalletProvider.tsx (hunk 3 above)
Comment (brendan-defi) [reply src/wallet/components/WalletProvider.tsx:46]: would an easier fix here be to add a parameter to handleClose called `animationDuration`?

File: tailwind.config.js (hunk 6 above)
Comment (brendan-defi) [reply tailwind.config.js:67]: i actually think i was overly specific about the name here, i think we should remove `wallet` and `walletIsland` from all the names, as they can be used for any container / container items.

File: tailwind.config.js (hunk 7 above)
Comment (brendan-defi) [reply tailwind.config.js:100]: see comment above, i think we should just make the names more generic as they are valid animations for any container / container element

File: tailwind.config.js (hunk 2 above)
Comment (brendan-defi) [reply tailwind.config.js:83]: v cool. just refactored all animations to use the new system

File: src/wallet/components/Wallet.test.tsx (hunk 1 above)
Comment (alessey) [reply src/wallet/components/Wallet.test.tsx:61]: is it possible to use a single <Wallet/> top level node and then look for WalletDropdown or WalletIsland to determine which?

File: src/wallet/components/WalletProvider.tsx [hunk 8]
```diff
@@ -15,11 +22,33 @@ export type WalletProviderReact = {
...
+  const handleClose = useCallback(() => {
+    if (!isOpen) {
+      return;
+    }
+
+    setIsClosing(true);
+    setTimeout(() => {
+      setIsOpen(false);
+      setIsClosing(false);
+    }, 300);
```
Comment (alessey) [inline src/wallet/components/WalletProvider.tsx:38]: you could watch the transitionEnd event instead of having to keep this delay and the animation in sync

File: src/wallet/components/WalletProvider.tsx (hunk 3 above)
Comment (dschlabach) [reply src/wallet/components/WalletProvider.tsx]: Hmm I think that'd still be suboptimal because it still requires you to keep the animation duration of the class and the prop in sync manually. If we use transitionEnd instead, then it'll offload that logic to the browser / React

File: src/styles/tailwind-base.css [hunk 9]
```diff
@@ -254,6 +254,16 @@
...
+  .animate-exitRight {
+    @apply animate-out fade-out slide-out-to-right-8 duration-500;
+  }
+
+  .animate-enterLeft {
+    @apply animate-in slide-in-from-left-8 fade-in duration-500;
+  }
+  .animate-exitLeft {
+    @apply animate-out fade-out slide-out-to-left-8 duration-500;
+  }
```
[the remark this answers is not in the document]
Comment (brendan-defi) [reply src/styles/tailwind-base.css:266]: ah, i see, i'll delete the classes and use the animation directives in-line

File: src/wallet/components/WalletProvider.tsx (hunk 3 above)
Comment (brendan-defi) [reply src/wallet/components/WalletProvider.tsx]: ack, i've refactored to use onAnimationEnd

File: src/wallet/components/WalletProvider.tsx (hunk 8 above)
Comment (brendan-defi) [reply src/wallet/components/WalletProvider.tsx]: i've refactored to remove the timeouts and use onAnimationEnd instead

File: src/wallet/components/Wallet.tsx [hunk 10]
```diff
@@ -1,23 +1,22 @@
...
 import { WalletProvider, useWalletContext } from './WalletProvider';
 
-const WalletContent = ({ children, className }: WalletReact) => {
-  const { isOpen, setIsOpen } = useWalletContext();
-  const walletContainerRef = useRef<HTMLDivElement>(null);
+function WalletContent({ children, className }: WalletReact) {
+  const {
+    isOpen,
+    handleClose,
+    containerRef: walletContainerRef,
```
Comment (alessey) [inline src/wallet/components/Wallet.tsx:16]: why did you need to move this ref up to the provider?

Comment (brendan-defi) [reply src/wallet/components/Wallet.tsx:16]: i actually don't anymore, i'll move this back down to the component

**Review by dschlabach — APPROVED** [review]

**Comment by alessey** [comment]
something weird going on with the dropdown close animation

Screen Recording 2024-12-18 at 10 33 36 AM

**Comment by cpcramer** [comment]
> something weird going on with the dropdown close animation
> 
> Screen Recording 2024-12-18 at 10 33 36 AM [ Screen Recording 2024-12-18 at 10 33 36 AM ](https://private-user-images.githubusercontent.com/2633933/396993040-4295b2b9-24e2-422b-ba3d-937fb1fbfc64.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ1NDU1ODksIm5iZiI6MTczNDU0NTI4OSwicGF0aCI6Ii8yNjMzOTMzLzM5Njk5MzA0MC00Mjk1YjJiOS0yNGUyLTQyMmItYmEzZC05MzdmYjFmYmZjNjQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMThUMTgwODA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Mjg3YTU2ZmVjM2VkMmJhNzg2YmY4MTZmYWZhN2NkNGYwNzE3YzdiMTZlOTg4OTY4MTA1YTZhMzZhYWIyNzBhYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Lydjb0cGwnjsUia8CvwIGQsWwZNN9HqaKwZX0mB2s6Y) [ ](https://private-user-images.githubusercontent.com/2633933/396993040-4295b2b9-24e2-422b-ba3d-937fb1fbfc64.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ1NDU1ODksIm5iZiI6MTczNDU0NTI4OSwicGF0aCI6Ii8yNjMzOTMzLzM5Njk5MzA0MC00Mjk1YjJiOS0yNGUyLTQyMmItYmEzZC05MzdmYjFmYmZjNjQuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMThUMTgwODA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Mjg3YTU2ZmVjM2VkMmJhNzg2YmY4MTZmYWZhN2NkNGYwNzE3YzdiMTZlOTg4OTY4MTA1YTZhMzZhYWIyNzBhYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Lydjb0cGwnjsUia8CvwIGQsWwZNN9HqaKwZX0mB2s6Y)

I'm also seeing this + a weird flicker when closing

**Comment by cpcramer** [comment]
The closing animation seems strange and feels slow 🤔

File: src/wallet/components/ConnectWallet.tsx [hunk 11]
```diff
@@ -63,8 +63,12 @@ export function ConnectWallet({
...
   // Handles
   const handleToggle = useCallback(() => {
-    setIsOpen(!isOpen);
-  }, [isOpen, setIsOpen]);
+    if (isOpen) {
+      handleClose();
+    } else {
+      setIsOpen(true);
+    }
+  }, [isOpen, handleClose, setIsOpen]);
```
Comment (cpcramer) [inline src/wallet/components/ConnectWallet.tsx:71]: Why do we need handleClose here? Is this for `WalletIsland`?

**Review by cpcramer — APPROVED** [review]

**Review by alessey — APPROVED** [review]
